internal/component/prometheus/remote/queue/network/loop.go

```diff
@@ -0,0 +1,374 @@
+package network
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/grafana/alloy/internal/component/prometheus/remote/queue/types"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/vladopajic/go-actor/actor"
+	"go.uber.org/atomic"
+)
+
+var _ actor.Worker = (*loop)(nil)
+
+// loop handles the low level sending of data. It conceptually a queue.
+// loop makes no attempt to save or restore signals in the queue.
+// loop config cannot be updated, it is easier to recreate. This does mean we lose any signals in the queue.
+type loop struct {
+	isMeta         bool
+	seriesMbx      actor.Mailbox[*types.TimeSeriesBinary]
+	client         *http.Client
+	batchCount     int
+	flushFrequency time.Duration
+	cfg            types.ConnectionConfig
+	log            log.Logger
+	lastSend       time.Time
+	statsFunc      func(s types.NetworkStats)
+	stopCalled     atomic.Bool
+	externalLabels map[string]string
+	series         []*types.TimeSeriesBinary
+	self           actor.Actor
+	ticker         *time.Ticker
+	req            *prompb.WriteRequest
+	buf            *proto.Buffer
+	sendBuffer     []byte
+}
+
+func newLoop(cc types.ConnectionConfig, isMetaData bool, log log.Logger, stats func(s types.NetworkStats)) *loop {
+	// TODO @mattdurham add TLS support afer the initial push.
+	l := &loop{
+		isMeta: isMetaData,
+		// In general we want a healthy queue of items, in this case we want to have 2x our maximum send sized ready.
+		seriesMbx:      actor.NewMailbox[*types.TimeSeriesBinary](actor.OptCapacity(2 * cc.BatchCount)),
+		batchCount:     cc.BatchCount,
+		flushFrequency: cc.FlushFrequency,
+		client:         &http.Client{},
+		cfg:            cc,
+		log:            log,
+		statsFunc:      stats,
+		externalLabels: cc.ExternalLabels,
+		ticker:         time.NewTicker(1 * time.Second),
+		buf:            proto.NewBuffer(nil),
+		sendBuffer:     make([]byte, 0),
+	}
+	l.req = &prompb.WriteRequest{
+		// We know BatchCount is the most we will ever send.
+		Timeseries: make([]prompb.TimeSeries, 0, cc.BatchCount),
+	}
+
+	return l
+}
+
+func (l *loop) Start() {
+	l.self = actor.Combine(l.actors()...).Build()
+	l.self.Start()
+}
+
+func (l *loop) Stop() {
+	l.stopCalled.Store(true)
+	l.self.Stop()
+}
+
+func (l *loop) actors() []actor.Actor {
+	return []actor.Actor{
+		actor.New(l),
+		l.seriesMbx,
+	}
+}
+
+func (l *loop) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Main select loop
+	select {
+	case <-ctx.Done():
+		l.stopCalled.Store(true)
+		return actor.WorkerEnd
+	// Ticker is to ensure the flush timer is called.
+	case <-l.ticker.C:
+		if len(l.series) == 0 {
+			return actor.WorkerContinue
+		}
+		if time.Since(l.lastSend) > l.flushFrequency {
+			l.trySend(ctx)
+		}
+		return actor.WorkerContinue
```

Comment:
Worth calling out that the smallest effective`flushFrequency` is 1s, because `ticker` is 1s.

Reply:
I will note it in the master PR component documentation that will be in the next PR.